daemon/src/db.rs

```diff
@@ -350,6 +549,393 @@ impl Connection {
+                let dlc_vout = Vout::new(vout);
+
+                self.lock = Some(Lock { txid, dlc_vout });
+
+                self.own_script_pubkey = Some(dlc.script_pubkey_for(self.role));
+            }
+            ContractSetupFailed => {}
+            OfferRejected => {}
+            RolloverStarted => {}
+            RolloverAccepted => {}
+            RolloverRejected => {}
+            RolloverCompleted { dlc, funding_fee } => {
+                self.own_script_pubkey = Some(dlc.script_pubkey_for(self.role));
+
+                self.fee_account.add_funding_fee(funding_fee);
+            }
+            RolloverFailed => {}
+            CollaborativeSettlementStarted { .. } => {}
+            CollaborativeSettlementProposalAccepted => {}
+            CollaborativeSettlementCompleted {
+                spend_tx,
+                script,
+                price,
+            } => {
+                let OutPoint { txid, vout } = spend_tx
+                    .outpoint(&script)
+                    .expect("spend script to be in spend tx");
+
+                let payout = &spend_tx.output[vout as usize];
+                let payout = Payout::new(Amount::from_sat(payout.value));
+
+                let txid = Txid::new(txid);
+                let vout = Vout::new(vout);
+
+                self.collaborative_settlement = Some(CollaborativeSettlement {
+                    txid,
+                    vout,
+                    payout,
+                    price,
+                })
+            }
+            CollaborativeSettlementRejected => {}
+            CollaborativeSettlementFailed => {}
+            LockConfirmed => {}
+            LockConfirmedAfterFinality => {}
+            CommitConfirmed => {}
+            CetConfirmed => {}
+            RefundConfirmed => {}
+            RevokeConfirmed => {}
+            CollaborativeSettlementConfirmed => {}
+            CetTimelockExpiredPriorOracleAttestation => {}
+            CetTimelockExpiredPostOracleAttestation { cet: _ } => {
+                // if we have an attestation we have already updated
+                // the `self.non_collaborative_settlement` field in
+                // the `OracleAttestedPriorCetTimelock` branch.
+                //
+                // We could repeat that work here just in case, but we
+                // don't have the closing price, so the
+                // `NonCollaborativeSettlement` struct would be
+                // incomplete
+            }
+            RefundTimelockExpired { refund_tx } => {
+                let own_script_pubkey = self
+                    .own_script_pubkey
+                    .as_ref()
+                    .expect("DLC to exist if refund timelock has expired");
+                let OutPoint { txid, vout } = refund_tx
+                    .outpoint(own_script_pubkey)
+                    .expect("spend script to be in spend tx");
+
+                let payout = &refund_tx.output[vout as usize];
```

Comment:
It is highly unlikely, but I've tried to remove all possible panics from `apply` by making it fallible. If we fail to apply the events we will get a handled error rather than a panic now.